### PR TITLE
Check for call context in `PartialFunction.infer_call_result`

### PR DESCRIPTION
**Pull request by nickdrozd**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|

## Description

Check a context's call context. Fixes four Mypy errors.

### CODE REVIEW

**Review by DanielNoord — COMMENTED** [review]
Shouldn't this be an `assert`? We're inferring a `call` so `callcontext` should probably be set

**Comment by DanielNoord** [comment]
I don't know, but it makes no sense to change the logic if we are inadvertently introducing bugs with it. This changes the behaviour of the function and feels a bit like moving [Chesterton's Fence](https://thoughtbot.com/blog/chestertons-fence).
I think a deepdive into `infer_call_result` would be really worthwhile but it should probably start from the top rather than doing individual changes in single definitions of it.

File: astroid/objects.py [hunk 1]
```diff
@@ -300,6 +300,7 @@ def __init__(
 
     def infer_call_result(self, caller=None, context: InferenceContext | None = None):
         if context:
+            assert context.callcontext
```
Comment (DanielNoord) [inline astroid/objects.py:303]:
```suggestion
            assert context.callcontext, "CallContext should be set before inferring call result"
```

**Review by DanielNoord — APPROVED** [review]

**Comment by DanielNoord** [comment]
I have just cleaned up the signature of `infer_call_result`. However, if not had any insight into whether `callcontext` is always set.

**Comment by DanielNoord** [comment]
I don't know. Asserting might very well be correct! Or we might want a subclass where the attribute has always been set?

**Comment by jacobtylerwalls** [comment]
I think this change makes sense for now. Whether CallContext must be part of the context object depends on whether or not the call is to a property. Here, this PR is only altering `PartialFunction.infer_call_result`, in a clear case where the code would have crashed if the assumption CallContext was always set was violated.

Here is where the CallContext is usually set if you are going through `infer_call`:

https://github.com/pylint-dev/astroid/blob/92fe8294690a3f83eaadaa63c3ec5cbd5ca8c90a/astroid/inference.py#L291-L295

Let's just get the mypy errors fixed 👍

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by DanielNoord** [comment]
> Let's just get the mypy errors fixed +1

Although I heavily agree with this I'm also wary of adding technical debt by adding asserts and code that we don't fully understand. In this case it might be fine but I do think we should be re-evaluating on a case by case basis and not just add them everywhere.

**Comment by jacobtylerwalls** [comment]
Agree! (I walked myself through the infer_call_result patterns before approving 📚 .)

**Pull request merged**